### PR TITLE
Use provider specific settings

### PR DESCRIPTION
**Pull request by kobelb**
The approach which we originally implemented where we extended from `AuthenticationProviderOptions` wasn't playing nicely with TypeScript. Long-term, we should allow auth providers to specify their config schema and ideally use something like [kbn-config-schema](https://github.com/elastic/kibana/tree/master/packages/kbn-config-schema) to generate types from these. However, we're not quite there yet, so we're stuck with `type AuthenticationProviderSpecificOptions = Record<string, unknown>` for the time being and doing the type checking in the auth provider constructor itself (even though the schema itself should prevent this from ever being needed).

I also made some changes to make the `xpack.security.authc.oidc.realm` required using `Joi` when `xpack.security.authProviders` contains `'oidc'`.

### CODE REVIEW

**Comment by jkakavas** [comment]
Thanks @kobelb ! LGTM, but I'll give Larry the chance to go through it too before applying

File: x-pack/plugins/security/server/lib/authentication/providers/oidc.ts [hunk 1]
```diff
@@ -113,11 +111,22 @@ function isAccessTokenExpiredError(err?: any) {
...
+    protected readonly options: Readonly<AuthenticationProviderOptions>,
+    oidcOptions: Readonly<AuthenticationProviderSpecificOptions>
+  ) {
     super(options);
-    if (!this.options.realm) {
+    if (!oidcOptions.realm) {
       throw new Error('Realm name must be specified');
     }
+
+    if (type(oidcOptions.realm) !== 'string') {
```
Comment (legrego) [inline x-pack/plugins/security/server/lib/authentication/providers/oidc.ts:125]: Not a criticism, just a question: what does this give us over the built in typeof operator?

**Review by legrego — APPROVED** [review]
LGTM!

File: x-pack/plugins/security/server/lib/authentication/providers/oidc.ts (hunk 1 above)
Comment (kobelb) [reply x-pack/plugins/security/server/lib/authentication/providers/oidc.ts:125]: For string type checking, we could just use typeof directly, it mainly comes in handy for those “gotcha types”.

**Pull request merged**